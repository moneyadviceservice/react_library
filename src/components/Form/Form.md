### A modified Column component that renders a form HTLM element. 
All props enabled by the Column component can be used here as well. This component also sets some default styles used by its children elements such as `input`.

```jsx
import { Container, Row, Form, Formfield, Select, Button } from '@moneypensionservice/directories';

<Container>
  <Row>
    <Form sizes={4}>
      <Formfield padding="0">
        <Select
          id="location"
          label="Location"
          options={['NYC', 'DC', 'LA', 'SF']} />
      </Formfield>
      <Button text="Submit" type="submit" />
    </Form>
  </Row>
</Container>
```
