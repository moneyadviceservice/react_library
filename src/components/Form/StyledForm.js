import styled, { css } from 'styled-components'
import { Col } from '../Grid'

const inputFocus = css`
  background-color: white;
  border-color: transparent;
`

const inputFocusOutline = css`
  outline: solid 0.1875rem ${props => props.theme.colors.form.inputFocus};
`

const StyledForm = styled(Col)`
  input[type='text'],
  input[type='number'],
  input[type='email'],
  input[type='search'],
  input[type='password'] {
    -webkit-appearance: none;
    border: 1px solid ${props => props.theme.colors.form.inputBorder};
    border-radius: 0;
    color: ${props => props.theme.colors.black};
    display: inline-block;
    height: 1.875rem;
    line-height: 1.875rem;

    &[disabled] {
      background-color: ${props => props.theme.colors.form.disabledBackground};
      color: ${props => props.theme.colors.form.disabledText};
    }
  }

  input[type='search'] {
    box-sizing: border-box;
  }

  input[type='text'],
  input[type='number'],
  input[type='email'],
  input[type='search'],
  input[type='password'],
  input[type='checkbox'],
  input[type='radio'],
  textarea {
    &:focus {
      ${inputFocus}
      ${inputFocusOutline}
    }
  }

  input[type='checkbox'],
  input[type='radio'] {
    &:hover {
      ${inputFocus}
      ${inputFocusOutline}
      cursor: pointer;
    }
  }

  input[type='search'] {
    &:focus {
      outline-offset: -1px;
    }
  }

  input[type='submit'],
  button,
  select {
    &:focus {
      ${inputFocusOutline}
    }
  }
`

export default StyledForm
