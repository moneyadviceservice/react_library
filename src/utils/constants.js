export const breakpoints = {
  xs: 1,
  sm: 48, // 768px
  md: 64, // 1024px
  lg: 90, // 1440px
  xl: 120, // 1920px
}

export const dimensions = ['xs', 'sm', 'md', 'lg', 'xl']

export const mediaQuery = 'only screen'

export const ALIGN_SELF_MAP = {
  center: 'center',
  end: 'flex-end',
  start: 'flex-start',
  stretch: 'stretch',
}

export const SPACING_SIZES = [
  'xxsmall',
  'xsmall',
  'small',
  'medium',
  'large',
  'xlarge',
]

