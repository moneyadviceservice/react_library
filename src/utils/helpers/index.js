export * from './background'
export * from './responsive'
export * from './flexbox'
export * from './styles'
export * from './mediaQueries'
export * from './sizing'
export * from './refs'
