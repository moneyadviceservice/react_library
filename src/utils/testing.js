import { render, act } from '@testing-library/react'
import GlobalTheme from '../components/ThemeProvider'

const customRender = (ui, options) =>
  render(ui, { wrapper: GlobalTheme, ...options })

const waitRender = async (ui, options) => {
  const { container, ...rest } = customRender(ui, options)

  await act(async () => {
    await Promise.resolve(container)
  })

  return { container, ...rest }
}

export * from '@testing-library/react'

export { customRender as render, waitRender }
