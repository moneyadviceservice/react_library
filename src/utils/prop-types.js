import PropTypes from 'prop-types'
import { SPACING_SIZES } from './constants'

/** spacing prop.type */
const spacingProp = PropTypes.shape({
  bottom: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
  horizontal: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
  left: PropTypes.oneOfType([PropTypes.oneOf(SPACING_SIZES), PropTypes.string]),
  right: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
  top: PropTypes.oneOfType([PropTypes.oneOf(SPACING_SIZES), PropTypes.string]),
  vertical: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
})

export const genericPropTypes = {
  a11yTitle: PropTypes.string,
  alignSelf: PropTypes.oneOf(['start', 'center', 'end', 'stretch']),
  margin: PropTypes.oneOfType([
    PropTypes.oneOf(['none', ...SPACING_SIZES]),
    spacingProp,
    PropTypes.string,
    PropTypes.object,
  ]),
  padding: PropTypes.oneOfType([
    PropTypes.oneOf([...SPACING_SIZES]),
    spacingProp,
    PropTypes.string,
    PropTypes.object,
  ]),
  border: PropTypes.oneOfType([
    PropTypes.oneOf(['none', ...SPACING_SIZES]),
    spacingProp,
    PropTypes.string,
    PropTypes.object,
  ]),
}

export function genericPropsDefaults({
  a11yTitle = null,
  alignSelf = null,
  margin = null,
  padding = null,
  border = null,
} = {}) {
  return {
    a11yTitle: a11yTitle,
    alignSelf: alignSelf,
    margin: margin,
    padding: padding,
    border: border,
  }
}
