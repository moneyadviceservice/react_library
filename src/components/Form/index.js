import React from 'react'
import PropTypes from 'prop-types'
import { genericPropTypes, genericPropsDefaults } from '../../utils/prop-types'
import StyledForm from './StyledForm'

const Form = ({ a11yTitle, as, children, ...rest }) => {
  return (
    <StyledForm aria-label={a11yTitle} forwardedAs={as} {...rest}>
      {children}
    </StyledForm>
  )
}

// Documentation
Form.propTypes = {
  /** Content inside component. */
  children: PropTypes.node,
  ...genericPropTypes,
}

Form.defaultProps = {
  as: 'form',
  ...genericPropsDefaults(),
}

/** @component */
export { Form }
