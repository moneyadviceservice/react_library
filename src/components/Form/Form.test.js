import React from 'react'
import { render } from '../../utils/testing'
import { Form } from '../Form'

describe('<Form />', () => {
  describe('Snapshots', () => {
    it('should match snapshot with children element', () => {
      const { container } = render(
        <Form>
          <p>example child</p>
        </Form>
      )
      expect(container).toMatchSnapshot()
    })
  })
})
