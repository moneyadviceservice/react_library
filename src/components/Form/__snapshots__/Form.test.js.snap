// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> Snapshots should match snapshot with children element 1`] = `
<div>
  <form
    class="styledCol__ColWrapper-sc-1mivnjm-0 fhguMU StyledForm-j6ft8m-0 hhgDcV"
  >
    <p>
      example child
    </p>
  </form>
</div>
`;
